acvm-repo/acvm_js/src/execute.rs

```diff
@@ -56,69 +68,174 @@ pub async fn execute_circuit(
+
+    execute_program_with_black_box_solver(&solver, program, initial_witness, &foreign_call_handler)
+        .await
+}
+
+#[wasm_bindgen(js_name = executeProgramWithBlackBoxSolver, skip_jsdoc)]
+pub async fn execute_program_with_black_box_solver(
+    solver: &WasmBlackBoxFunctionSolver,
+    program: Vec<u8>,
+    initial_witness: JsWitnessMap,
+    foreign_call_executor: &ForeignCallHandler,
+) -> Result<JsWitnessStack, Error> {
+    let witness_stack = execute_program_with_native_type_return(
+        solver,
+        program,
+        initial_witness,
+        foreign_call_executor,
+    )
+    .await?;
+
+    Ok(witness_stack.into())
+}
+
+async fn execute_program_with_native_type_return(
+    solver: &WasmBlackBoxFunctionSolver,
+    program: Vec<u8>,
+    initial_witness: JsWitnessMap,
+    foreign_call_executor: &ForeignCallHandler,
+) -> Result<WitnessStack, Error> {
     let program: Program = Program::deserialize_program(&program)
-        .map_err(|_| JsExecutionError::new("Failed to deserialize circuit. This is likely due to differing serialization formats between ACVM_JS and your compiler".to_string(), None))?;
-    let circuit = match program.functions.len() {
-        0 => return Ok(initial_witness),
-        1 => &program.functions[0],
-        _ => return Err(JsExecutionError::new("Program contains multiple circuits however ACVM currently only supports programs containing a single circuit".to_string(), None).into())
-    };
-
-    let mut acvm = ACVM::new(&solver.0, &circuit.opcodes, initial_witness.into());
-
-    loop {
-        let solver_status = acvm.solve();
-
-        match solver_status {
-            ACVMStatus::Solved => break,
-            ACVMStatus::InProgress => {
-                unreachable!("Execution should not stop while in `InProgress` state.")
-            }
-            ACVMStatus::Failure(error) => {
-                let (assert_message, call_stack) = match &error {
-                    OpcodeResolutionError::UnsatisfiedConstrain {
-                        opcode_location: ErrorLocation::Resolved(opcode_location),
-                    }
-                    | OpcodeResolutionError::IndexOutOfBounds {
-                        opcode_location: ErrorLocation::Resolved(opcode_location),
-                        ..
-                    } => {
-                        (circuit.get_assert_message(*opcode_location), Some(vec![*opcode_location]))
+    .map_err(|_| JsExecutionError::new("Failed to deserialize circuit. This is likely due to differing serialization formats between ACVM_JS and your compiler".to_string(), None))?;
+
+    let main = &program.functions[0];
+
+    let executor = ProgramExecutor::new(&program.functions, &solver.0, &foreign_call_executor);
+
+    let mut witness_stack = WitnessStack::default();
+    let main_witness =
+        executor.execute_circuit(main, initial_witness.into(), &mut witness_stack).await?;
+    witness_stack.push(0, main_witness);
```

Comment:
This seems like a strange interface. Will look at in more detail later.

Reply:
Do you mean with passing the mutable witness stack or attaching the main witness at the end? 

This isn't too far off what we do here (https://github.com/noir-lang/noir/blob/c5a7242f6d43eb3a7e74995a6334e38cc3c45ba5/tooling/nargo/src/ops/execute.rs#L151) as well as other places in the SSA where we have special handling for main. It just made things easier for async execution if I had `execute_circuit` accept an immutable reference to self.

Reply:
I see what you had in mind now. Going to resolve this and can make a followup for the nargo op.